### ISSUE TITLE
Potential training bugs: repeat_sequence truncation and MAPPO agent-ID mismatch after shuffling

### ISSUE THREAD

**Issue by Credica** [TTomilin/MEAL#11]
Hi, thank you for releasing MEAL. While auditing the current training code for continual MARL experiments, I noticed two independent issues that may affect repeated-task plasticity experiments and MAPPO training.

I checked commit [8a9fff5](https://github.com/TTomilin/MEAL/commit/8a9fff5502c5318a2872db2d57d29fade97a1b71) on the current <code>main</code> branch with a clean worktree.

## 1. repeat_sequence creates repeated environments, but only seq_length tasks are trained

The sequence loader expands the base task list:

~~~python
full_specs = base_specs * max(1, int(repeat_sequence))
~~~

See [sequence_loader.py#L126-L131](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/meal/env/generation/sequence_loader.py#L126-L131).

However, the IPPO training loop creates RNGs using the unrepeated <code>seq_length</code> and then zips them with the expanded environment list:

~~~python
rng, *env_rngs = jax.random.split(rng, seq_length + 1)

for task_idx, (rng, env) in enumerate(zip(env_rngs, envs)):
    ...
~~~

See [ippo.py#L962-L979](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/ippo.py#L962-L979).

The same pattern appears in:

- [mappo.py#L838-L844](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/mappo.py#L838-L844)
- [happo.py#L1141-L1149](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/happo.py#L1141-L1149)

Because Python <code>zip</code> stops at the shorter iterable, a configuration such as:

~~~text
seq_length = 10
repeat_sequence = 3
~~~

creates 30 environments but trains only the first 10. The repeated portion of the stream is therefore never visited.

This appears to be a regression. When <code>repeat_sequence</code> was integrated in commit [4be7d43](https://github.com/TTomilin/MEAL/commit/4be7d433), the loop used:

~~~python
jax.random.split(rng, len(envs) + 1)
~~~

Commit [802b059](https://github.com/TTomilin/MEAL/commit/802b0594) later changed <code>len(envs)</code> to <code>seq_length</code> in the RNG split, network task/head count, evaluation count, and AGEM task slots.

### Expected behavior

All <code>seq_length * repeat_sequence</code> task occurrences should be processed.

### Actual behavior

Only <code>seq_length</code> task occurrences are processed.

### Semantic question

It may be useful to clarify whether a repeated occurrence should:

1. reuse the original semantic task ID and output head, or
2. receive a new occurrence-specific task ID and output head.

Changing only the RNG split to <code>len(envs)</code> may not be sufficient, because the network heads, evaluation arrays, and CL memory are currently also sized by <code>seq_length</code>.

## 2. MAPPO recomputes agent IDs from shuffled minibatch positions

MAPPO enables agent-ID conditioning by default:

~~~python
use_agent_id: bool = True
~~~

The actor derives each ID from the current input batch position:

~~~python
batch_size = x.shape[0]
agent_ids = jnp.arange(batch_size) // self.num_envs
agent_onehot = jax.nn.one_hot(agent_ids, self.num_agents)
~~~

See [decoupled_mlp.py#L89-L97](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/model/decoupled_mlp.py#L89-L97).

This is correct during rollout, where <code>batchify</code> produces the blocked ordering:

~~~text
agent0_env0 ... agent0_envN
agent1_env0 ... agent1_envN
...
~~~

During PPO optimization, however, the time/actor dimensions are flattened and randomly permuted before minibatches are constructed:

~~~python
batch = jax.tree_util.tree_map(
    lambda x: x.reshape((batch_size,) + x.shape[2:]),
    batch,
)
permutation = jax.random.permutation(_rng, batch_size)
shuffled_batch = jax.tree_util.tree_map(
    lambda x: jnp.take(x, permutation, axis=0),
    batch,
)
~~~

See [mappo.py#L683-L697](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/mappo.py#L683-L697).

The actor is then called on each shuffled minibatch without an explicit agent ID:

~~~python
pi, _ = actor_network.apply(
    params["actor"], local_obs, env_idx=env_idx
)
~~~

See [mappo.py#L522-L528](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/mappo.py#L522-L528).

Consequently, the ID inferred from the new row position no longer corresponds to the sample's original agent.

With the current defaults:

~~~text
num_envs        = 2048
num_agents      = 2
num_steps       = 400
num_minibatches = 16

flattened batch size = 2048 * 2 * 400 = 1,638,400
minibatch size       = 1,638,400 / 16 = 102,400
derived agent IDs    = arange(102,400) // 2048 = 0,...,49
valid agent IDs      = 0,1
~~~

Therefore, 98,304 of 102,400 rows (96%) receive an out-of-range ID and are encoded as an all-zero one-hot vector. JAX documents that out-of-range indices passed to <code>jax.nn.one_hot</code> are encoded as zeros: [JAX documentation](https://docs.jax.dev/en/latest/_autosummary/jax.nn.one_hot.html).

The remaining 4% are also not reliably labeled, because their assigned ID is determined by their post-shuffle row position rather than by the originating agent.

### Expected behavior

The true agent identity should remain associated with each transition through flattening, permutation, minibatching, and replay.

### Actual behavior

The actor reconstructs agent identity from the minibatch row index after the original ordering has been destroyed.

### Possible direction

One option would be to store <code>agent_id</code> explicitly in the transition and apply the same reshape/permutation operations to it as to observations, actions, and advantages. The actor could then receive the preserved ID directly rather than infer it from batch position.

A regression test could attach known IDs to a small multi-agent batch, shuffle it with a fixed permutation, and verify that each observation remains paired with its original ID.

## Potential impact

The first issue may prevent repetition-based plasticity experiments from actually reaching the repeated tasks on the current code path. The second may remove or corrupt MAPPO's intended agent-identity conditioning during optimization while leaving rollout-time conditioning apparently correct.

Please let me know if I have misunderstood the intended semantics of <code>repeat_sequence</code> or the minibatch layout used by the MAPPO actor.
